### PR TITLE
Rlecomte/tdi 43679 pubsub output

### PR DESCRIPTION
**Pull request by rlecomte-talend**
https://jira.talendforge.org/browse/TDI-43679

### CODE REVIEW

File: common-stream-io/stream-avro/src/main/java/org/talend/components/common/stream/output/avro/HeadlessAvroRecordWriter.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.EncoderFactory;
+import org.talend.components.common.stream.api.output.RecordConverter;
+import org.talend.components.common.stream.api.output.RecordWriter;
+import org.talend.components.common.stream.api.output.TargetFinder;
+import org.talend.sdk.component.api.record.Record;
+
+import java.io.IOException;
+
+public class HeadlessAvroRecordWriter implements RecordWriter {
```
Comment (undx) [inline common-stream-io/stream-avro/src/main/java/org/talend/components/common/stream/output/avro/HeadlessAvroRecordWriter.java]: Maybe give a different  name  for this writer...  or maybe make a <> constructor in AvroRecordWriter...

File: pubsub/src/main/java/org/talend/components/pubsub/output/message/CSVMessageGenerator.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class CSVMessageGenerator extends MessageGenerator {
+
+    private char fieldDelimiter;
+
+    protected static class CSVRecordVisitor implements RecordVisitor<String> {
```
Comment (undx) [inline pubsub/src/main/java/org/talend/components/pubsub/output/message/CSVMessageGenerator.java]: Maybe sharing this visitor to common-x might be a good idea?

File: pubsub/src/test/java/org/talend/components/pubsub/output/message/CSVMessageGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                .withEntry(
+                        rbf.newEntryBuilder().withName("nullableString").withType(Schema.Type.STRING).withNullable(true).build())
+
+                .build();
+        Record record = rbf.newRecordBuilder(schema).withString("name", "John Smith").withInt("age", 42).withString("state", "CA")
+                .build();
+
+        PubsubMessage message = beanUnderTest.generateMessage(record);
+        Assertions.assertNotNull(message, "Message is null");
+        System.out.println(message.getData().toStringUtf8());
```
Comment (undx) [inline pubsub/src/test/java/org/talend/components/pubsub/output/message/CSVMessageGeneratorTest.java]: no sout please.

File: package.json [hunk 4]
```diff
@@ -5,7 +5,7 @@
   "author": "Anas Khabali <akhabali@talend.com>",
   "license": "Apache-2.0",
   "dependencies": {
-    "@talend/i18n-scripts": "0.7.0"
+    "@talend/i18n-scripts": "0.7.1"
```
[the remark this answers is not in the document]
Comment (rlecomte-talend) [reply package.json]: corrected

File: azure/azureblob/pom.xml [hunk 5]
```diff
@@ -35,7 +35,7 @@
 
         TIP: it is recommended to set this property into your settings.xml in an active by default profile.
         -->
-        <talend.component.studioHome />
+        <talend.component.studioHome/>
```
[the remark this answers is not in the document]
Comment (rlecomte-talend) [reply azure/azureblob/pom.xml]: change reverted

File: common-stream-io/common-io/src/main/java/org/talend/components/common/stream/format/avro/AvroConfiguration.java [hunk 6]
```diff
@@ -12,8 +12,14 @@
  */
 package org.talend.components.common.stream.format.avro;
 
+import lombok.Data;
 import org.talend.components.common.stream.format.ContentFormat;
 
+@Data
 public class AvroConfiguration implements ContentFormat {
 
+    private boolean attachSchema = true;
```
[the remark this answers is not in the document]
Comment (rlecomte-talend) [reply common-stream-io/common-io/src/main/java/org/talend/components/common/stream/format/avro/AvroConfiguration.java]: Let's live discuss about this

File: pubsub/src/test/java/org/talend/components/pubsub/output/message/CSVMessageGeneratorTest.java (hunk 3 above)
Comment (rlecomte-talend) [reply pubsub/src/test/java/org/talend/components/pubsub/output/message/CSVMessageGeneratorTest.java]: removed sysout

File: pubsub/src/main/java/org/talend/components/pubsub/output/message/CSVMessageGenerator.java (hunk 2 above)
Comment (rlecomte-talend) [reply pubsub/src/main/java/org/talend/components/pubsub/output/message/CSVMessageGenerator.java]: RecordVistor is already part of TCK and this implementation doesn't do anything really useful pour other connectors. Future CSV writer should use the stream-csv component.

File: pubsub/src/main/java/org/talend/components/pubsub/output/PubSubOutput.java [hunk 7]
```diff
@@ -70,14 +70,9 @@ public void onElement(Record record) {
...
         init = true;
-        if (configuration.getTopicOperation() == PubSubOutputConfiguration.TopicOperation.CREATE_IF_NOT_EXISTS) {
-            pubSubService.createTopicIfNeeded(configuration.getDataset().getDataStore(), configuration.getDataset().getTopic());
-        } else {
-            Topic topic = pubSubService.loadTopic(configuration.getDataset().getDataStore(),
-                    configuration.getDataset().getTopic());
-            if (topic == null) {
-                throw new PubSubConnectorException(i18n.topicDoesNotExist(configuration.getDataset().getTopic()));
-            }
+        Topic topic = pubSubService.loadTopic(configuration.getDataset().getDataStore(), configuration.getDataset().getTopic());
```
Comment (clesaec) [inline pubsub/src/main/java/org/talend/components/pubsub/output/PubSubOutput.java]: May add local var for topic & datastore ?

File: pubsub/src/main/resources/org/talend/components/pubsub/service/Messages.properties [hunk 8]
```diff
@@ -18,6 +18,12 @@ org.talend.components.pubsub.service.I18nMessage.errorReadCSV = Cannot read CSV
 org.talend.components.pubsub.service.I18nMessage.errorCredentials = Cannot create credentials for Google Cloud: {0}.
 org.talend.components.pubsub.service.I18nMessage.errorRemoveSubscription = Cannot remove subscription: {0}.
 org.talend.components.pubsub.service.I18nMessage.errorCreateSubscription = Cannot create subscription: {0}.
+org.talend.components.pubsub.service.I18nMessage.errorCreateTopic = Cannot create topic: {0}.
+org.talend.components.pubsub.service.I18nMessage.errorRemoveTopic = Cannot remove topic: {0}.
+org.talend.components.pubsub.service.I18nMessage.errorWriteCSV = An error occured while writing CSV message: {0}.
+org.talend.components.pubsub.service.I18nMessage.errorWriteAvro = An error occured while wrting AVRO message: {0}.
```
Comment (fdesiles) [inline pubsub/src/main/resources/org/talend/components/pubsub/service/Messages.properties]: "i" missing in Writing.

**Pull request merged**